lib/Bio/KBase/ObjectAPI/KBaseFBA/FBA.pm

```diff
@@ -2564,34 +2569,53 @@ sub setupFBAExperiments {
 			$phenoko =~ s/\|/___/g;
 			push(@{$phenoData},$pheno->id()."\t".$phenoko."\t".$media."\t".$pheno->normalizedGrowth());
 		}
-		foreach my $key (keys(%{$mediaHash})) {
-			push(@{$medialist},$mediaHash->{$key});
-		}
-		Bio::KBase::ObjectAPI::utilities::PRINTFILE($self->jobDirectory()."/".$fbaExpFile,$phenoData);
-	} elsif (length($self->mediaset_ref()) > 0 || @{$self->media_list()} > 0) {
-		$self->parameters()->{"phenotype analysis"} = 1;
-		$self->parameters()->{"save phenotype simulation fluxes"} = 1;
-		$fbaExpFile = "FBAExperiment.txt";
-		my $mediaHash = {};
-		if (length($self->mediaset_ref()) > 0) {
-			my $mediaset = $self->mediaset();
-			for (my $i=0; $i < @{$mediaset->{elements}}; $i++) {
-				push(@{$self->media_list_refs()},$mediaset->{elements}->[$i]->{"ref"});
-			}
-		}
-		my $inmedialist = $self->media_list();
-		for (my $i=0; $i < @{$inmedialist}; $i++) {
-			$mediaHash->{$inmedialist->[$i]->name()} = $inmedialist->[$i];
-			push(@{$phenoData},$inmedialist->[$i]->name()."\tnone\t".$inmedialist->[$i]->name()."\t1");
-		}
-		foreach my $key (keys(%{$mediaHash})) {
-			push(@{$medialist},$mediaHash->{$key});
-		}
-		Bio::KBase::ObjectAPI::utilities::PRINTFILE($self->jobDirectory()."/".$fbaExpFile,$phenoData);
+
+        push @$medialist, values %$mediaHash;
+
+        Bio::KBase::ObjectAPI::utilities::PRINTFILE(
+            $self->jobDirectory() . "/" . $fbaExpFile,
+            $phenoData
+        );
+
+        return $fbaExpFile;
+
 	}
+
+    if ( $self->mediaset_ref && length( $self->mediaset_ref ) > 0
+        || @{ $self->media_list } ) {
+
+        $self->parameters()->{ "phenotype analysis" } = 1;
+        $self->parameters()->{ "save phenotype simulation fluxes" } = 1;
+        $fbaExpFile = "FBAExperiment.txt";
+
+        if ( $self->mediaset_ref && length( $self->mediaset_ref ) ) {
+            my $mediaset = $self->mediaset();
+            if ( $mediaset->{ elements } && @{ $mediaset->{ elements } } ) {
+                push @{ $self->media_list_refs() },
+                    map { $_->{ 'ref' } } @{ $mediaset->{ elements } };
+            }
+        }
+
+        if ( @{ $self->media_list } ) {
+            my %mediaHash;
+            for my $medium ( @{ $self->media_list } ) {
+                push @$phenoData, join "\t",
+                    $medium->name, "none", $medium->name, 1;
+                $mediaHash{ $medium->name } = $medium;
+            }
+            push @$medialist, values %mediaHash;
```

Comment:
unless `$self->media_list` contains duplicates, using `%mediaHash` is unnecessary -- it could be replaced with 
```
push @$medialist, @{ $self->media_list };
```
If `$self->media_list` does contain duplicates, they will be copied into `$phenoData`.